fedimint-client/src/db.rs

```diff
@@ -262,3 +276,221 @@ impl_db_lookup!(
     key = ClientModuleRecovery,
     query_prefix = ClientModuleRecoveryPrefix
 );
+
+/// `ClientMigrationFn` is a function that modules can implement to "migrate"
+/// the database to the next database version.
+pub type ClientMigrationFn = for<'r, 'tx> fn(
+    &'r mut DatabaseTransaction<'tx>,
+    ModuleInstanceId,
+    Vec<DynState>, // active states
+    Vec<DynState>, // inactive states
```

Comment:
In order for the module to do be able to determine which state the raw bytes are, the decoder still needs to be registered with the db right? If we can't decode the old format I don't think the migration can do anything interesting (maybe I'm missing a use case here)

Reply:
dev call: using bytes here allows the migration to change the fields of a struct. Old struct needs to still be kept around so it can be decoded, but it can be removed from the state machine enum

Reply:
Unless I have a misunderstanding, I think passing `DynState` here is more convenient for module developers, since they can always call `to_bytes` on the state, decode it using `consensus_decode`, and do the migration with the resulting struct (which doesn't need to be a state listed in the module's state machine).

Reply:
Ah nevermind, states that have been deleting can no longer be encoded as `DynState` so this will not work. `Vec<u8>` is the only option, working on a test for this now

Reply:
Fixed. Changed this to `Vec<u8>` and changed the test to be as close to a state deletion as possible.